[WPE] Platform: Fix crash on Wayland in DMA-BUF feedback handling

aperezdc opened the pull request:
#### a3f9ca459d73096f2da94be8ca54d1b514f268eb
<pre>
[WPE] Platform: Fix crash on Wayland in DMA-BUF feedback handling
<a href="https://bugs.webkit.org/show_bug.cgi?id=273178">https://bugs.webkit.org/show_bug.cgi?id=273178</a>

Reviewed by Carlos Garcia Campos.

Move creation of the pending DMABufFeedback into the &quot;main_device&quot;
handler, which according to the protocol should be the first one
in the sequence. Also, move reusing the previous FormatTable into the
&quot;done&quot; event callback, because at that point it is certain that there
will not be more related events coming.

The rest of the changes are needed to be able to create empty
FormatTable objects for the initial creation of new DMABufFeedback
instance, and being able move-assign it later.

* Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
(DMABufFeedback::FormatTable::FormatTable):
(DMABufFeedback::FormatTable::operator bool const):
(DMABufFeedback::FormatTable::operator=):
(DMABufFeedback::DMABufFeedback):
(DMABufFeedback::format):
(wpeViewWaylandGetPreferredDMABufFormats):
(DMABufFeedback::create): Deleted.

Canonical link: <a href="https://commits.webkit.org/278091@main">https://commits.webkit.org/278091@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/009240e794f7cff3e02f0eefad7ef709854b5ba0

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/49461 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/28743 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/52497 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/52700 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/134 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/34760 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/26363 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/40378 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/51561 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/26289 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/42616 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/21496 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/23747 "Passed tests") | | [✅ 🛠 wpe-skia](https://ews-build.webkit.org/#/builders/52/builds/7830 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/45663 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/44304 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/54212 "Built successfully") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/24543 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/20733 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/47745 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/25815 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/8/builds/42812 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/46759 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/26654 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/7108 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/25538 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -83,6 +83,29 @@ struct DMABufFeedback {
+            std::swap(size, other.size);
+            std::swap(data, other.data);
+            return *this;
+        }
```
This is inconsistent with the move constructor. We can change the move constructor to just do: `*this = WTFMove(other);` so that if this changes in the future we only need to change the operator here.

aperezdc replied:
Good idea, I'll apply the simplification.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -93,13 +116,9 @@ struct DMABufFeedback {
-            return nullptr;
-
-        return makeUnique<DMABufFeedback>(FormatTable(size, data));
+        return makeUnique<DMABufFeedback>(FormatTable());
```
I guess we can now change the DMABufFeedback constructor to not receive a FormatTable, since we always want an empty one now.

aperezdc replied:
Good idea.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -83,6 +83,29 @@ struct DMABufFeedback {
+            : size(size)
+            , data(static_cast<FormatTable::Data*>(mmap(nullptr, size, PROT_READ, MAP_PRIVATE, fd, 0)))
+        {
+            if (data == MAP_FAILED) {
+                data = nullptr;
+                size = 0;
+            }
```
This is problematic, because we assume that a DMABufFeedback can't be created with an invalid format table (DMABufFeedback::create() fails if mmap fails). For example in `DMABufFeedback::format()` we use the format table without checking it's valid.

aperezdc replied:
I understand that the compositor must at least once give a valid `format_table` event during the lifetime of the listener, but indeed it won't hurt to have some check in place before trying to use invalid data.

Another question: should we try to fall-back to read the table from the file descriptor if `mmap()` fails? I suppose it's reasonable to expect that the file descriptor mapping will work given that we should be always receive a valid file descriptor in the event callback... Or does anyone feel differently about this?

carlosgcampos replied:
I'm not sure it's worth it, why would mmap fail but read don't?

aperezdc replied:
Let's assume for now that reading the file descriptor would fail also if `mmap()` failed. I tend to think it's probably not worth it to handle, becase mapping would fail if we're running out of virtual address space, and at that point we would have bigger problems.

aperezdc replied:
I've added bounds checking in the `format()` method, including a debug assertion because compositors should never reference an invalid format index from a tranche, and they have to provide a format table. There's some graceful handling for release builds.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -406,6 +425,11 @@ static const struct zwp_linux_dmabuf_feedback_v1_listener linuxDMABufFeedbackLis
+        // Compositor might not have sent the format table. In that case, try to reuse the previous one.
+        if (!view->priv->pendingDMABufFeedback->formatTable && view->priv->committedDMABufFeedback && view->priv->committedDMABufFeedback->formatTable)
+            view->priv->pendingDMABufFeedback->formatTable = WTFMove(view->priv->committedDMABufFeedback->formatTable);
+
```
We need to handle here the case of format table being invalid and set pending feedback to nullptr in that case. Because in current code we know that if we have a committed feedback, it has a valid table.

aperezdc replied:
Right. If we end up without any usable formats table at the `done` callback (either because one wasn't sent, or we don't have a previously committed one) then we need to bail out early and avoid sending the signal as well. Good catch!

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -414,16 +438,22 @@ static const struct zwp_linux_dmabuf_feedback_v1_listener linuxDMABufFeedbackLis
+        //      Is there any compositor out there which does not conform to the
+        //      protocol spec?
+        if (!priv->pendingDMABufFeedback)
+            priv->pendingDMABufFeedback = DMABufFeedback::create();
```
I don't see where in the spec it says that main device is first. In any case we are already handling this case, but maybe we can remove the comment.

aperezdc replied:
You can check it [here, for example](https://wayland.app/protocols/linux-dmabuf-v1#zwp_linux_dmabuf_feedback_v1) (that site pulls the data from the XML protocol specification files):

> To send parameters, the compositor sends one main_device event, tranches (each consisting of one tranche_target_device event, one tranche_flags event, tranche_formats events and then a tranche_done event), then one done event.

carlosgcampos replied:
Right, that doesn't say if format table goes after or before main device

aperezdc replied:
I'm going to remove most of the comment, including the `XXX` marker (usually I use those for myself) and leave a small note like “The protocol specification is not clear about the ordering of the format_table and main_device events, so allow both to err on the safe side.»

carlosgcampos reviewed (approved):
Looks good to me, but we need some changes before landing.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -83,6 +80,29 @@ struct DMABufFeedback {
+        FormatTable& operator=(FormatTable&& other)
+        {
+            std::swap(size, other.size);
+            std::swap(data, other.data);
```
Why swap? what we want here is that the moved object is reset after the move. That will always happen for the move constructor, but for the assignment operator, the object might have a value that we don't want in the moved object. I think std::exchange is what we want here, or am I missing something?

aperezdc replied:
If `this->data` was valid we need to `munmap()` it. The swap is there to make the possibly valid values of `data` and `size` end up in `other`: the destructor will be called on `other` after the function returns because the compiler knows `other` cannot be used further because it was moved into the function call. The destructor on `other` will call `munmap()` for the previous value of `this->data`. I understand this is a common RAII idiom.

carlosgcampos replied:
Then, let's do the unmap here (but only if the map is different, of course). We need to make sure that after committed = WTFMove(pending); pending is reset, we don't want to swap them unless committed is already in a reset state.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -93,17 +113,12 @@ struct DMABufFeedback {
-            return nullptr;
-
-        return makeUnique<DMABufFeedback>(FormatTable(size, data));
+        return makeUnique<DMABufFeedback>();
```
Thinking about this again, I guess we could just call makeUnique directly, the create() doesn't add anything in this case.

aperezdc replied:
Right, I will remove it before landing.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -93,17 +113,12 @@ struct DMABufFeedback {
 
-    explicit DMABufFeedback(FormatTable&& table)
-        : formatTable(WTFMove(table))
+    explicit DMABufFeedback()
```
explicit is no longer needed

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -406,24 +425,40 @@ static const struct zwp_linux_dmabuf_feedback_v1_listener linuxDMABufFeedbackLis
+
+        // The compositor might not have sent the formats table. In that case, try to reuse the previous
+        // one. Return early and skip emitting the signal if there is no usable formats table in the end.
+        if (!view->priv->pendingDMABufFeedback->formatTable) {
```
```suggestion
        if (view->priv->pendingDMABufFeedback && !view->priv->pendingDMABufFeedback->formatTable) {
```

aperezdc replied:
Wouldn't it rather be “if there's no pending feedback object, return early”?

aperezdc replied:
Ah, I see you suggest the early return below 👌🏼

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -406,24 +425,40 @@ static const struct zwp_linux_dmabuf_feedback_v1_listener linuxDMABufFeedbackLis
+            if (view->priv->committedDMABufFeedback && view->priv->committedDMABufFeedback->formatTable)
+                view->priv->pendingDMABufFeedback->formatTable = WTFMove(view->priv->committedDMABufFeedback->formatTable);
+            else {
+                view->priv->pendingDMABufFeedback.reset();
+                return;
```
This is a change in behaviour, but I think it's a good idea. Maybe we should early return if we don't have a pending feedback too, it's a compositor bug for sure, but we already have early returns from all other callbacks.

aperezdc replied:
Yes, I thought it would be weird that we would store as “commited” a feedback instance without any formats and produce the signal&mdash;it would be unusable.

As you point out, arriving to the `done` event without a valid `priv->pendingDMABufFeedback` would be a compositor bug, because it must have been created either in the `main_device` or the `format_table` event callbacks. The [protocol spec](https://wayland.app/protocols/linux-dmabuf-v1#zwp_linux_dmabuf_feedback_v1) is ambiguous about the ordering between `main_device` and `format_table`, but it is very clear about `main_device` being always present:

> The parameters are sent once when this object is created and whenever they change. The done event is always sent once after all parameters have been sent. When a single parameter changes, all parameters are re-sent by the compositor.
> [...]
> To send parameters, the compositor sends one main_device event, tranches (each consisting of one tranche_target_device event, one tranche_flags event, tranche_formats events and then a tranche_done event), then one done event.

This means that the checks on `priv->pendingDMABufFeedback` in each of the `tranche_*` event callbacks are not needed in theory. Let's leave them for now as a safeguard.

carlosgcampos commented on Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
```diff
@@ -795,7 +830,8 @@ static WPEBufferDMABufFormats* wpeViewWaylandGetPreferredDMABufFormats(WPEView*
         for (const auto& format : tranche.formats) {
             auto [fourcc, modifier] = priv->committedDMABufFeedback->format(format);
-            wpe_buffer_dma_buf_formats_builder_append_format(builder, fourcc, modifier);
+            if (LIKELY(fourcc != 0))
```
```suggestion
            if (LIKELY(fourcc))
```

aperezdc committed:
[WPE] Platform: Fix crash on Wayland in DMA-BUF feedback handling
https://bugs.webkit.org/show_bug.cgi?id=273178

Reviewed by Carlos Garcia Campos.

Move creation of the pending DMABufFeedback into the "main_device"
handler, which according to the protocol should be the first one
in the sequence. Also, move reusing the previous FormatTable into the
"done" event callback, because at that point it is certain that there
will not be more related events coming.

The rest of the changes are needed to be able to create empty
FormatTable objects for the initial creation of new DMABufFeedback
instance, and being able move-assign it later.

* Source/WebKit/WPEPlatform/wpe/wayland/WPEViewWayland.cpp:
(DMABufFeedback::FormatTable::FormatTable):
(DMABufFeedback::FormatTable::operator bool const):
(DMABufFeedback::FormatTable::operator=):
(DMABufFeedback::DMABufFeedback):
(DMABufFeedback::format):
(wpeViewWaylandGetPreferredDMABufFormats):
(DMABufFeedback::create): Deleted.

Canonical link: https://commits.webkit.org/278091@main
